### DISCUSSION TITLE
The great Log update

### DISCUSSION THREAD

**Discussion by tadhgboyle** [NamelessMC/Nameless#2310] [Ideas]
Logging in NamelessMC is a little used feature, but has a lot of potential.

As you may know, the Log class is very annoying currently.
You have to use the `Action` function to get the "action", but... what does this even do? No language translations exist so its just a very vague `snake_case_phrase` which is non translatable.

My proposed solution: 
- Still use the Action array, but change it's layout:
    ```php
   private const ACTION = [
      'api' => [
         'discord_role_set' => [
            'file' => 'api',
            'term' => 'discord_role_set',
            'category' => 'discord'
         ]
      ]
   ]
   ```
   - Layout: 
      - "api": category of this action
      - "discord_role_set": specific action
         - "file": language file to look for the "term" in, which will display on the Security logs list page
         - "category": which specific log category this action will be under, may be the same or different as the root category ("api" in this case"). If not defined, will use root category

**NOT DONE YET**

**Comment by Derkades** [comment]
I'm not sure translating log messages is a good idea.

* Translating is a lot more work and very few people are going to look at log messages. Probably just developers who need to be familiar with English anyway.
* Consistent error messages allow easy searching in discord or github for answers
* Logging should be simple, involving more components allows it to break more easily

EDIT: I may have misinterpreted the issue, is this about audit logging or standard debug/info/warn/error log messages? Translations do make sense for audit logging.

**Comment by tadhgboyle** [comment]
Yeah, its audit logging

**Comment by samerton** [comment]
Thanks for raising this, it's something that's been on my mind for years.

Something I want to add to the system is before/after values wherever possible, so we can revert a change with a single click.

**Comment by tadhgboyle** [comment]
Yeah I was thinking of that too, not sure the best way to store it all though. Any smart ideas?
The log table will need two more columns, `before` and `after`, then I guess we make a script to get the action from each log and then it would know what to do to revert to the `before` state. Not sure if that makes sense. Absolutely open to ideas

**Comment by partydragen** [comment]
well like i did on NamelessHosting v3 to change most deletions to be a deleted colum and its timestamp instead of 1 then somehow add that id into log to allow to restore and its sub stuff

**Comment by partydragen** [comment]
But what about other modules? We must use own action const?
